source/slang/slang-ir-variable-scope-correction.cpp

```diff
@@ -0,0 +1,307 @@
+            }
+        }
+    }
+}
+
+// Check if the instruction is used outside of the loop.
+// The loopHeaderList contains all the loop headers where the original instruction is defined.
+// So we if the block of the user instruction is dominated by the break block of the loop header,
+// it means that it was out of the loop, so it's out of the scope of the loop.
+// Note the reason we use the loopHeaderList is because there could be nested loops, so we need to
+// check all the loop headers from inner to outer.
+bool VariableScopeCorrectionContext::_isOutOfScopeUse(IRInst * userInst, IRDominatorTree* domTree, const List<IRLoop*>& loopHeaderList)
+{
+    if (auto block = getBlock(userInst))
+    {
+        // If the use site of this instruction is dominated by the break block, it means that the
+        // instruction is used after the break block, so we need to make that instruction available globally.
+        // By doing so, we record all the users of this instructions.
+        for(auto loopHeader : loopHeaderList)
+        {
+            auto breakBlock = loopHeader->getBreakBlock();
+            if (domTree->dominates(breakBlock, block))
+            {
+                return true;
+            }
+        }
+    }
+    return false;
+}
+
+IRInst* VariableScopeCorrectionContext::_processInstruction(IRDominatorTree* dominatorTree, IRInst* instAfterParam,
+        IRInst* originInst, IRInst* useInst, const List<IRLoop*>& loopHeaderList, const HashSet<IRInst*> workList)
+{
+    // For a given instruction, we need to check all the users of this instruction to see
+    // if the user is out of the scope of the loop. If so, we need to make the instruction available globally.
+    // In addition, we have to recursively check all the operands of this instruction to see if they are also
+    // out of the scope of the loop.
+
+    // Check if the user of this instruction is out of the scope of the loop
+    if(_isOutOfScopeUse(useInst, dominatorTree, loopHeaderList))
+    {
+        bool isStoable = _isStorableType(originInst->getDataType());
+        IRInst* newInst = nullptr;
+        if(isStoable)
+        {
+            newInst = _processStorableInst(instAfterParam, originInst, useInst);
+        }
+        else
+        {
+            // duplicate the instruction right before the use site.
+            newInst = _processUnstorableInst(originInst, useInst);
+            // When we duplicate the out-of-scope instruction, the side-effect is that the operands of the instruction
+            // could also be out of the scope. So we have to recursively check all the operands of the instruction. Such
+            // issue dooesn't happen for the storable instruction because we store the result of the instruction in a new
+            // variable visible in the whole function.
+            UInt operandCount = newInst->getOperandCount();
```

Comment:
This can be irBuilder.replaceOperand(use, newInst)

The remaining of the code here is not necessary. Basically you want processInst to make a single use valid at a time.